### PR TITLE
eventloop: aggregate endpointdiscoveries into one

### PR DESCRIPTION
**Pull request by ilackarms**
this pr has two features:
- aggregate all the different endpointdiscoveries into one watcher
- refactor the plugin registry and the way we call the endpoint discovery init functions (we now have a plugin type that implements the SetupEndpointDiscovery function)

based on #75

### CODE REVIEW

**Comment by yuval-k** [comment]
it seems that the test are failing for some missing interface?

**Comment by yuval-k** [comment]
@ilackarms can you rebase off master?

**Comment by ilackarms** [comment]
@yuval-k test passed and rebased. please re-review

File: internal/control-plane/endpointswatcher/endpoints_watcher.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+			}
+		}(ed)
+	}
+}
+
+// start a goroutine that takes every new endpoints group received from an ed
+// and aggregates them to a single set of EndpointGroups which get passed to our own channel
+func (e *endpointsAggregator) aggregateReceivedEndpoints(wg *sync.WaitGroup, stop <-chan struct{}) {
+	// watch eds for new sets of endpoints
+	endpointsByDiscovery := make(map[endpointdiscovery.Interface]endpointdiscovery.EndpointGroups)
```
Comment (yuval-k) [inline internal/control-plane/endpointswatcher/endpoints_watcher.go]: should this be inside the go-routine?

Comment (ilackarms) [reply internal/control-plane/endpointswatcher/endpoints_watcher.go]: does it make any difference? i can move it there, not sure it matters

File: internal/control-plane/endpointswatcher/endpoints_watcher.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+func (e *endpointsAggregator) TrackUpstreams(upstreams []*v1.Upstream) {
+	wg := &sync.WaitGroup{}
+	for _, discovery := range e.endpointDiscoveries {
+		wg.Add(1)
+		go func(epd endpointdiscovery.Interface) {
+			defer wg.Done()
+			epd.TrackUpstreams(upstreams)
+		}(discovery)
+	}
+	wg.Wait()
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply internal/control-plane/endpointswatcher/endpoints_watcher.go]: `TrackUpstreams` is intended to block (it starts a sync)

**Pull request merged**